File: 1. Arrays/1. Two Sum/two-sum-optimal.py
from typing import List


def twoSum(nums: List[int], target: int):
    n = len(nums)
    mp = {}
    for i in range(0, n):
        number_to_find = target - nums[i]
        if number_to_find in mp:
            return [mp[number_to_find], i]
        else:
            mp[nums[i]] = i
    return [-1, -1]


print(twoSum([2, 7, 11, 15], 18))

File: 1. Arrays/1. Two Sum/two-sum-optimal.cpp
#include <iostream>
#include <bits/stdc++.h>

using namespace std;

class Solution
{
public:
    vector<int> twoSum(vector<int> &nums, int target)
    {
        unordered_map<int, int> mp;
        for (int i = 0; i < nums.size(); i++)
        {
            int numToFind = target - nums[i];
            if(mp.find(numToFind) != mp.end()){
                return vector<int>{mp[numToFind],i};
            }else{
                mp[nums[i]] = i;
            }
        }
        return vector<int>{-1,-1};
    }
};

int main()
{
    return 0;
}